standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/DirectSqlInsertPart.java

```diff
@@ -0,0 +1,804 @@
+  }
+
+  private void insertPartitionInBatch(Map<Long, MPartition> partIdToPartition, Map<Long, Long> partIdToSdId)
+      throws MetaException {
+    int rowCount = partIdToPartition.size();
+    String columns = "(\"PART_ID\",\"CREATE_TIME\",\"LAST_ACCESS_TIME\",\"PART_NAME\",\"SD_ID\",\"TBL_ID\","
+        + "\"WRITE_ID\")";
+    int columnCount = 7;
+    BatchExecutionContext batchExecutionContext = new BatchExecutionContext() {
+      final Iterator<Map.Entry<Long, MPartition>> it = partIdToPartition.entrySet().iterator();
+      @Override
+      public void execute(String batchQueryText, int batchRowCount, int batchParamCount) throws MetaException {
+        Object[] params = new Object[batchParamCount];
+        int paramIndex = 0;
+        for (int index = 0; index < batchRowCount; index++) {
+          Map.Entry<Long, MPartition> entry = it.next();
+          MPartition partition = entry.getValue();
+          params[paramIndex++] = entry.getKey();
+          params[paramIndex++] = partition.getCreateTime();
+          params[paramIndex++] = partition.getLastAccessTime();
+          params[paramIndex++] = partition.getPartitionName();
+          params[paramIndex++] = partIdToSdId.get(entry.getKey());
+          params[paramIndex++] = partition.getTable().getId();
+          params[paramIndex++] = partition.getWriteId();
+        }
+        executeQuery(batchQueryText, params);
+      }
+    };
+    insertInBatch("\"PARTITIONS\"", columns, columnCount, rowCount, batchExecutionContext);
+  }
+
+  private void insertSerdeParamInBatch(Map<Long, MSerDeInfo> serdeIdToSerDeInfo) throws MetaException {
+    int rowCount = 0;
+    for (MSerDeInfo serDeInfo : serdeIdToSerDeInfo.values()) {
+      rowCount += serDeInfo.getParameters().size();
+    }
+    if (rowCount == 0) {
+      return;
+    }
+    String columns = "(\"SERDE_ID\",\"PARAM_KEY\",\"PARAM_VALUE\")";
+    int columnCount = 3;
+    BatchExecutionContext batchExecutionContext = new BatchExecutionContext() {
+      final Iterator<Map.Entry<Long, MSerDeInfo>> serdeIt = serdeIdToSerDeInfo.entrySet().iterator();
+      Map.Entry<Long, MSerDeInfo> serdeEntry = serdeIt.next();
+      Iterator<Map.Entry<String, String>> it = serdeEntry.getValue().getParameters().entrySet().iterator();
+      @Override
+      public void execute(String batchQueryText, int batchRowCount, int batchParamCount) throws MetaException {
+        Object[] params = new Object[batchParamCount];
+        int index = 0;
+        int paramIndex = 0;
+        do {
+          while (index < batchRowCount && it.hasNext()) {
+            Map.Entry<String, String> entry = it.next();
+            params[paramIndex++] = serdeEntry.getKey();
+            params[paramIndex++] = entry.getKey();
+            params[paramIndex++] = entry.getValue();
+            index++;
+          }
+          if (index < batchRowCount) {
+            serdeEntry = serdeIt.next();
```

Comment:
Should check `serdeIt.hasNext()` here and see if the serdeEntry.getValue().getParameters() is null?

Reply:
serdeIt.next() is accessed only if have any more rows to insert. I mean, it is already ensured not to access null. And serdeEntry.getValue().getParameters() cannot be null. It is either empty list or list with values.

Have added comment now to make it clear.

Reply:
Nit: the `skewedColNames`, `skewedColValues` and `skewedColValueLocationMaps` may be null in MStorageDescriptor:
https://github.com/apache/hive/blob/master/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java#L2526-L2542

Reply:
done